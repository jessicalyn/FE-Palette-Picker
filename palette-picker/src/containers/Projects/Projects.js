import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchOptionsCreator } from '../../utils/fetchOptionsCreator';
import { postNewProject } from '../../thunks/postNewProject';

export class Projects extends Component {
  constructor() {
    super();
    this.state = {
      project_name: '',
      project_id: 0
    }
  }

  updateProjectName = (e) => {
    const { name, value } = e.target
    this.setState({ [name]: value })
  }
  
  addNewProject = async (e) => {
    e.preventDefault()
    const projectName = this.state.project_name
    const body = { project_name: projectName }
    const options = await fetchOptionsCreator('POST', body)
    this.props.postNewProject(options, projectName)
  }

  render() {
    const projectsToDisplay = this.props.projects.map(project => {
      return <h3>{project.project_name}</h3>
    })
    return(
      <div>
      <form onSubmit={this.addNewProject}>
        <input type="text" onChange={this.updateProjectName} value={this.state.project_name} name="project_name" />
        <button>Add New Project</button>
      </form>
      {projectsToDisplay}
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  projects: state.projects
})

export const mapDispatchToProps = (dispatch) => ({
  postNewProject: (url, body) => dispatch(postNewProject(url, body))
})

export default connect(mapStateToProps, mapDispatchToProps)(Projects)