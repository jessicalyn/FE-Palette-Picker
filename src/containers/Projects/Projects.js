import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchOptionsCreator } from '../../utils/fetchOptionsCreator';
import { postNewProject } from '../../thunks/postNewProject';
import closeButton from '../../assets/closeButton.png';
import { storePalettes, storeProjects } from '../../actions';
import { deletePalette } from '../../thunks/deletePalette';
import { deleteProject } from '../../thunks/deleteProject';

export class Projects extends Component {
  constructor() {
    super();
    this.state = {
      project_name: '',
      project_id: 0
    }
  }

  updateProjectName = (e) => {
    const { name, value } = e.target
    this.setState({ [name]: value })
  }
  
  addNewProject = async (e) => {
    e.preventDefault()
    const projectName = this.state.project_name
    const body = { project_name: projectName }
    const options = await fetchOptionsCreator('POST', body)
    this.props.postNewProject(options, projectName)
  }

  deletePalette = (e) => {
    const id = parseInt(e.target.id)
    const updatedPalettes = this.props.palettes.filter(palette => id !== palette.palette_id)
    this.props.storePalettes(updatedPalettes)
    this.props.deletePalette(id)
  }

  deleteProject = (e) => {
    const id = parseInt(e.target.id)
    const updatedProjects = this.props.projects.filter(project => id !== project.project_id)
    this.props.storeProjects(updatedProjects)
    this.props.deleteProject(id)
  }

  render() {
    const projectsToDisplay = this.props.projects.map((project, index) => {
      return <div>
        <h3 key={`${project.project_name}-${index}`}>{project.project_name}</h3>
          <button onClick={this.deleteProject}>
            <img src={closeButton} alt={'Delete Project icon'} id={project.project_id}/>
          </button>
        {
          this.props.palettes.map(palette => {
            if (palette.project_id === project.project_id) {
              return <div>
                <h4>{palette.palette_name}</h4>
                <button onClick={this.deletePalette}>
                  <img src={closeButton} alt={'Delete Palette icon'} id={palette.palette_id}/>
                </button>
              </div>
            }
          })
        }
      </div>
    })
    return(
      <div>
      <form onSubmit={this.addNewProject}>
        <input type="text" onChange={this.updateProjectName} value={this.state.project_name} name="project_name" />
        <button className="project-button">Add New Project</button>
      </form>
      {projectsToDisplay}
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  projects: state.projects,
  palettes: state.palettes
})

export const mapDispatchToProps = (dispatch) => ({
  postNewProject: (url, body) => dispatch(postNewProject(url, body)),
  storePalettes: (updatedPalettes) => dispatch(storePalettes(updatedPalettes)),
  deletePalette: (id) => dispatch(deletePalette(id)),
  storeProjects: (updatedProjects) => dispatch(storeProjects(updatedProjects)),
  deleteProject: (id) => dispatch(deleteProject(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(Projects)