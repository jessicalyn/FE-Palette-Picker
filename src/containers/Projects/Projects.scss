.projects-to-display {
  // display: flex;
  border: 2px solid purple
}

.individual-projects {
  display: flex;
  border: 2px solid lightgrey;
  background-color: whitesmoke;
  border-radius: 10px;
  margin: 15px;
}

.project-palettes {
  // display: flex;
  border-radius: 10px;
  border: 2px solid lightgrey;
  margin: 10px;
  width: 125px;
  height: 100px;
  background-color: white;
}

.palette-header {
  display: flex;
}

h5 {
  flex: 1;
}

.palette-header button {
  flex: 1;
  flex-direction: row-reverse;
}

.delete-palette-button {
  width: 20px;
  height: 20px;
}

.palette-colors {
  display: flex;
}

.small-palette {
  height: 50px;
  width: 50px;
  border: 1px solid black
}