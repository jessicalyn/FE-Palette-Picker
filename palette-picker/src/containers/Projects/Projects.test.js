import { Projects, mapStateToProps, mapDispatchToProps } from './Projects';
import React from 'react';
import { shallow } from 'enzyme';
import { fetchOptionsCreator } from '../../utils/fetchOptionsCreator';
jest.mock('../../utils/fetchOptionsCreator');

describe('Projects Container', () => {
  describe('Projects', () => {
    let wrapper;
    const mockState = {
      project_name: '',
      project_id: 0
    }
    const mockProjects = [
      { project_id: 1,
        project_name: "Travel App",
        created_at: "2019-05-10T17:45:09.612Z",
        updated_at: "2019-05-10T17:45:09.612Z"
      },
      { project_id: 2,
        project_name: "Fitness App",
        created_at: "2019-05-10T17:45:09.634Z",
        updated_at: "2019-05-10T17:45:09.634Z"
      }
    ]
    let mockPostNewProject;

    beforeEach(() => {
      mockPostNewProject = jest.fn()
      wrapper = shallow(
        <Projects 
          projects={mockProjects}
          postNewProject={mockPostNewProject}
        />
      )
    })

    it('should match the snapshot', () => {
      expect(wrapper.debug()).toMatchSnapshot()
    })

    it('should have proper default state', () => {
        expect(wrapper.state()).toEqual(mockState)
    })

    it('should set state with a project name', () => {
      const mockEvent = { target: { name : "project_name", value: "MyNewProject" } }
      
      wrapper.instance().updateProjectName(mockEvent)

      expect(wrapper.state()).toEqual({
        project_name: "MyNewProject",
        project_id: 0
      })
    })

    it('should call fetchOptionsCreator with a method and body', async () => {
      const mockEvent = { preventDefault: jest.fn() }
      const projectName = "Unique Project Name"

      wrapper.setState({
        project_name: projectName
      })

      wrapper.instance().addNewProject(mockEvent)

      expect(fetchOptionsCreator).toHaveBeenCalledWith('POST', { project_name: projectName} )
    })

    it.skip('should call postNewProject with options and projectName as arguments', () => {
      const mockEvent = { preventDefault: jest.fn() }
    })

  })

  describe('mapStateToProps', () => {
    const mockState = {
      projects: [
        { project_id: 1,
          project_name: "Travel App",
          created_at: "2019-05-10T17:45:09.612Z",
          updated_at: "2019-05-10T17:45:09.612Z"
        },
        { project_id: 2,
          project_name: "Fitness App",
          created_at: "2019-05-10T17:45:09.634Z",
          updated_at: "2019-05-10T17:45:09.634Z"
        }
      ],
      fakeState: "Not real state to return"
    }
    const expected = {
      projects: [
        { project_id: 1,
          project_name: "Travel App",
          created_at: "2019-05-10T17:45:09.612Z",
          updated_at: "2019-05-10T17:45:09.612Z"
        },
        { project_id: 2,
          project_name: "Fitness App",
          created_at: "2019-05-10T17:45:09.634Z",
          updated_at: "2019-05-10T17:45:09.634Z"
        }
      ]
    }

    const mappedProps = mapStateToProps(mockState)
    expect(mappedProps).toEqual(expected)
  })

  describe('mapDispatchToProps', () => {
    const mockUrl = 'www.projects.com'
    const mockBody = { project_name: "Harold's Project" }
    const mockDispatch = jest.fn()
    const postNewProject = jest.fn()
    const actionToDispatch = postNewProject(mockUrl, mockBody)
    const mappedProps = mapDispatchToProps(mockDispatch)

    mappedProps.postNewProject(mockUrl, mockBody)

    expect(mockDispatch).toHaveBeenCalled(actionToDispatch)
  })
  
})