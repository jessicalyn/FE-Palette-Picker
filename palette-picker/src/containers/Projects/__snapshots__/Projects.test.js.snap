// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects Container Projects should match the snapshot 1`] = `
"<div>
  <form onSubmit={[Function]}>
    <input type=\\"text\\" onChange={[Function]} value=\\"\\" name=\\"project_name\\" />
    <button>
      Add New Project
    </button>
  </form>
  <h3>
    Travel App
  </h3>
  <h3>
    Fitness App
  </h3>
</div>"
`;
